### ISSUE TITLE
Activity timeouts do not count against retry limits.

### ISSUE THREAD

**Issue by mark-r-cullen** [Azure/azure-functions-durable-extension#582]
We have an orchestrator that has a retry activity set to retry 3 times, with the function time out set to 10 minutes.

I could be wrong, but from what I can see so far it seems as though if the activity function times out and throws a FunctionTimeoutException, it is not being counted as a retry. It looks like it is just retrying the activity function forever, or perhaps until it doesn't time out.

Does this sound like expected behaviour?

**Comment by cgillum** [comment]
In the past (Functions 1.0), exceeding the function timeout caused the entire function app to shut down. That shutdown would not be detected as a failure, so it would also not get counted towards the number of retries. I don't recall if that behavior changed in Functions 2.0, however. What is the callstack of the FunctionTimeoutException you are observing?

**Comment by mark-r-cullen** [comment]
Hi,

I think this is what we are seeing with a v2.0 function app. When I was trying to debug the issue and the function timed out, the function CLI died completely and the debugger stopped...

I've attached the stack trace from app insights: [function timeout.txt](https://github.com/Azure/azure-functions-durable-extension/files/2783407/function.timeout.txt)

I guess this is probably not something that you are easily able to handle, so we will have to just work around this by throwing our own timeout exception, before the function times out?

Thanks for your help.

**Comment by cgillum** [comment]
Yes unfortunately we don't have a good way of automatically handling this right now. Implementing your own timeout mechanism should work as a workaround.

**Comment by ConnorMcMahon** [comment]
My most recent PR may have some implications for this: https://github.com/Azure/azure-functions-durable-extension/pull/1242.

In general, it is a very good idea to do whatever possible to prevent function timeouts, as this has implications on the entire instance that times out. 

That being said, maybe we have a way to properly count timeouts as failures.

**Comment by YunchuWang** [comment]
Closing this as fixed by #1917.

That change introduced a distinct `FunctionTimeoutError` result and routes an activity timeout through the same `TaskActivityShim` failure path as an unhandled user-code exception. The shim throws `TaskFailureException`, so DTFx records a failed activity attempt and `CallActivityWithRetry` consumes it against the configured retry limit instead of treating the timeout as a transient work-item abort and retrying indefinitely.

#1917 also added C# and Python timeout test apps that exercise a real Functions `functionTimeout` against a slow activity, and its release note explicitly records the activity/entity timeout handling fix. This resolves the host-reported function-timeout behavior described in this issue.
